### PR TITLE
Add a mode to avoid clone() in DDPSink

### PR DESCRIPTION
**Pull request by pritamdamania87**
DDPSink clones the outputs of DDP to avoid in-place modification of loss (see https://github.com/pytorch/pytorch/issues/61982). However, when outputs are really large (2-3GB) this adds a lot of overhead for peak memory.

As a result, adding a mode to avoid this clone in cases where users are not modifying loss in-place.

cc @mrshenli @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

### CODE REVIEW

**Review by fegin — APPROVED** [review]
LGTM

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -896,6 +898,9 @@ def __init__(
         if self._use_python_reducer:
             self._register_accum_grad_hook()
 
+        # Whether or not DDPSink performs a clone.
+        self.ddp_sink_clone = True
```
Comment (rohan-varma) [inline torch/nn/parallel/distributed.py:902]: nit: maybe make this attribute a private one

**Review by rohan-varma — APPROVED** [review]
Makes sense, thanks!

**Comment by pritamdamania87** [comment]
Thanks for the review @fegin and @rohan-varma!

**Comment by pritamdamania87** [comment]
@fegin @rohan-varma Could you approve the appropriate workflows as suggested in https://github.com/pytorch/pytorch/pull/122927#issuecomment-2043566328? Thanks!

**Pull request closed without merging**